server/src/test/java/io/crate/integrationtests/SnapshotRestoreIntegrationTest.java

```diff
@@ -976,6 +981,65 @@ public void test_restore_partitioned_tables_rename_all() throws Exception {
         );
     }
 
+    @Test
+    public void test_restore_old_users() throws IOException {
+        File repoDir = TEMPORARY_FOLDER.getRoot().toPath().toAbsolutePath().toFile();
+        try (InputStream stream = Files.newInputStream(getDataPath("/repos/oldusersmetadata_repo.zip"))) {
+            TestUtil.unzip(stream, repoDir.toPath());
+        }
+        execute(
+            "CREATE REPOSITORY users_repo TYPE \"fs\" with (location=?, compress=true, readonly=true)",
+            new Object[]{repoDir.getAbsolutePath()}
+        );
+        execute("CREATE USER \"John\" WITH (password='johns-password')");
+        execute("CREATE USER \"Arthur\"");
+        execute("CREATE USER \"Marios\"");
+        execute("SELECT * FROM sys.users ORDER BY name");
+        assertThat(response).hasRows(
+            "Arthur| NULL| false",
+            "John| ********| false",
+            "Marios| NULL| false",
+            "crate| NULL| true");
+
+        // Snapshot contains the following users:
+        // CREATE USER "Arthur" WITH (password='arthurs-password');
+        // CREATE USER "Ford" WITH (password='fords-password');
+        // CREATE USER "John";
+        execute("RESTORE SNAPSHOT users_repo.usersnap USERS with (wait_for_completion=true)");
+
+        execute("SELECT * FROM sys.users ORDER BY name");
+        assertThat(response).hasRows(
+            "Arthur| ********| false",
+            "Ford| ********| false",
+            "John| NULL| false",
+            "crate| NULL| true");
+
+        // Before any CREATE/ALTER/DROP operation RolesMetadata still have the users&roles defined
+        // but only UsersMetadata are used
+        RolesMetadata rolesMetadata = cluster().clusterService().state().metadata().custom(RolesMetadata.TYPE);
+        assertThat(rolesMetadata).isNotNull();
+        Assertions.assertThat(rolesMetadata.roles()).containsOnlyKeys("Arthur", "John", "Marios");
+        UsersMetadata usersMetadata = cluster().clusterService().state().metadata().custom(UsersMetadata.TYPE);
+        assertThat(usersMetadata).isNotNull();
+        assertThat(usersMetadata.users()).containsOnlyKeys("Arthur", "Ford", "John");
+
+        execute("ALTER USER \"Arthur\" SET (password='arthurs-new-password')");
+        execute("SELECT * FROM sys.users ORDER BY name");
+        assertThat(response).hasRows(
+            "Arthur| ********| false",
+            "Ford| ********| false",
+            "John| NULL| false",
+            "crate| NULL| true");
```

Comment:
With ALTER + SELECT we are testing that we can change password of the restored user, right?
If so, I'm not sure this assertion does its job: output is exactly the same as output of previous SELECT and password is masked. 

Can we store prev password and here select only password of Arthur and check old != new?

Reply:
The test is mainly to show that the ALTER USER is dropping UsersMetadata (and existing RolesMetadata) and creates a new RolesMetadata out of this restored UsersMetadata.

I've changed the stmt to change the password for user `John`.